App a11y assert api

nishath-bs opened the pull request:
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Polish (an improvement for an existing feature)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (improvements to the project's docs)
- [ ] Specification changes (updates to WebDriver command specifications)
- [ ] Internal updates (everything related to internal scripts, governance documentation and CI files)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [ ] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Backport Request

[//]: # (The current `main` branch is the development branch for WebdriverIO v9. If your change should be released to the current major version of WebdriverIO (v8), please raise another PR with the same changes against the `v8` branch.)

- [ ] This change is solely for `v9` and doesn't need to be back-ported
- [ ] Back-ported PR at `#XXXXX`

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

nishath-bs committed:
performScan changes for app-a11y

wdio-bot committed:
v8.41.0

harsha509 committed:
catch failures onAfterCommand (#13993)

dbs-2001 committed:
add neccessary variables for a11y in env

dbs-2001 committed:
change polling timeout env

dbs-2001 committed:
resolve error

dbs-2001 committed:
remove unnecessary changes

nishath-bs committed:
added polling api for app a11y results

nishath-bs committed:
Merge pull request #3 from dbs-2001/v8

V8

dbs-2001 committed:
checking os version even if given in regex form

kamal-kaur04 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -419,6 +427,24 @@ export const launchTestSession = o11yErrorHandler(async function launchTestSessi
     }
 })
 
+export const validateCapsWithAppA11y = (deviceName?: any, platformMeta?: { [key: string]: any; }) => {
```
do we need deviceName here?

nishath-bs replied:
Removed

kamal-kaur04 commented on packages/wdio-browserstack-service/src/accessibility-handler.ts:
```diff
@@ -93,23 +99,35 @@ class _AccessibilityHandler {
         }
 
+        if (isBrowserstackSession(this._browser) && isAppAccessibilityAutomationSession(this._accessibility, this.isAppAutomate)) {
+            const deviceName = this._getCapabilityValue(this._caps, 'deviceName', 'device')
```
we can remove this, we don't have any validation on deviceName

nishath-bs replied:
Done

kamal-kaur04 commented on packages/wdio-browserstack-service/src/accessibility-handler.ts:
```diff
@@ -216,8 +245,26 @@ class _AccessibilityHandler {
+            }
             const isPageOpened = await this.checkIfPageOpened(this._browser, uniqueId, shouldScanScenario)
-
+            if (!isPageOpened) {
```
reason for moving this?

nishath-bs replied:
AppAutomate session doesn't need this pageOpened check its applicable only for automate session

kamal-kaur04 replied:
Ya this isn't removed. This is moved later in the code.

kamal-kaur04 commented on packages/wdio-browserstack-service/src/accessibility-handler.ts:
```diff
@@ -287,16 +330,18 @@ class _AccessibilityHandler {
-        const results: unknown = await (browser as WebdriverIO.Browser).executeAsync(accessibilityScripts.saveTestResults as string, dataForExtension)
-        BStackLogger.debug(util.format(results as string))
+        await performA11yScan(this.isAppAutomate, browser, true, true)
+        if (!isAppAccessibilityAutomationSession(this._accessibility, this.isAppAutomate)) {
```
if (isAppAccessibilityAutomationSession) return;

kamal-kaur04 replied:
facing any issues in that?

nishath-bs replied:
Done

kamal-kaur04 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -483,7 +509,39 @@ export const isAccessibilityAutomationSession = (accessibilityFlag?: boolean | s
-export const performA11yScan = async (browser: WebdriverIO.Browser | WebdriverIO.MultiRemoteBrowser, isBrowserStackSession?: boolean, isAccessibility?: boolean | string, commandName?: string) : Promise<{ [key: string]: any; } | undefined> => {
+export const isAppAccessibilityAutomationSession = (accessibilityFlag?: boolean | string, isAppAutomate?: boolean) => {
+    try {
+        const hasA11yJwtToken = typeof process.env.BSTACK_A11Y_JWT === 'string' && process.env.BSTACK_A11Y_JWT.length > 0 && process.env.BSTACK_A11Y_JWT !== 'null' && process.env.BSTACK_A11Y_JWT !== 'undefined'
```
please move to a utility method... all the checks

nishath-bs replied:
I have reused `isAccessibilityAutomationSession` itself now. See if it is fine

kamal-kaur04 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -526,7 +589,67 @@ export const getA11yResults = async (browser: WebdriverIO.Browser, isBrowserStac
 
-export const getA11yResultsSummary = async (browser: WebdriverIO.Browser, isBrowserStackSession?: boolean, isAccessibility?: boolean | string) : Promise<{ [key: string]: any; }> => {
+export const getAppA11yResults = async (isAppAutomate: boolean, browser: WebdriverIO.Browser, isBrowserStackSession?: boolean, isAccessibility?: boolean | string, sessionId?: string | null) : Promise<Array<{ [key: string]: any; }>> => {
+    if (!isBrowserStackSession) {
```
didn't get this check

nishath-bs replied:
borrowed this from WebA11y code

kamal-kaur04 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -526,7 +589,67 @@ export const getA11yResults = async (browser: WebdriverIO.Browser, isBrowserStac
+        const upperTimeLimit = Date.now() + 360000 // 30 seconds
+        const params = { test_run_uuid: process.env.TEST_ANALYTICS_ID, session_id: sessionId, timestamp: Date.now() } // Query params to pass
+        const header = { Authorization: `Bearer ${process.env.BSTACK_A11Y_JWT}` }
+        console.log(`params ${JSON.stringify(params)}`)
```
please remove debug logs once done with dev. Add necessary logs using logger

nishath-bs replied:
Done

kamal-kaur04 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -526,7 +589,67 @@ export const getA11yResults = async (browser: WebdriverIO.Browser, isBrowserStac
+        const header = { Authorization: `Bearer ${process.env.BSTACK_A11Y_JWT}` }
+        console.log(`params ${JSON.stringify(params)}`)
+        console.log(`header ${JSON.stringify(header)}`)
+        const apiRespone = await pollApi(apiUrl, params, header, upperTimeLimit)
```
where are we using polling Timeout? Also, logic seems different as compared to TestNG in terms of buffer seconds we are taking.

nishath-bs replied:
What logic difference are there?

kamal-kaur04 replied:
I thought about this 360000 :) But this seems like a miss

nishath-bs replied:
Let's keep it as 30 secs

nishath-bs replied:
Changed

kamal-kaur04 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -1359,3 +1482,66 @@ export const isValidCapsForHealing = (caps: { [key: string]: Options.Testrunner
+            return pollApi(url, params, headers, upperLimit, startTime)
+        }
+        console.error('Unexpected error occurred:')
+        return { data: {}, headers: {}, message: 'Unexpected error occurred.' }
```
messages should be coming from backend right?

nishath-bs replied:
Handled

nishath-bs committed:
Merge pull request #4 from nishath-bs/fix-os-version-regex

checking os version even if given in regex form

nishath-bs committed:
review comment changes

kamal-kaur04 commented on packages/wdio-browserstack-service/src/accessibility-handler.ts:
```diff
@@ -149,7 +164,19 @@ class _AccessibilityHandler {
+                this._testMetadata[testIdentifier].accessibilityScanStarted = shouldScanTest
 
+                if (shouldScanTest) {
+                    BStackLogger.info('AppAutomate test case execution has started.')
```
Let's keep it Automate only.. same in TestNG

nishath-bs committed:
added instrumentation

kamal-kaur04 commented on packages/wdio-browserstack-service/src/accessibility-handler.ts:
```diff
@@ -216,8 +243,26 @@ class _AccessibilityHandler {
+                }
+                Listener.setTestRunAccessibilityVar(this._accessibility && shouldScanScenario)
+                if (shouldScanScenario) {
+                    BStackLogger.info('AppAutomate test case execution has started.')
```
same comment here

kamal-kaur04 commented on packages/wdio-browserstack-service/src/accessibility-handler.ts:
```diff
@@ -216,8 +243,26 @@ class _AccessibilityHandler {
+                }
+                return
+            }
             const isPageOpened = await this.checkIfPageOpened(this._browser, uniqueId, shouldScanScenario)
```
Can we add !isApp condition on isPageOpened only rather than duplicating the code.

More better, we can removed for web A11y as well not needed now `isPageOpened`?

+ @amaanbs

kamal-kaur04 commented on packages/wdio-browserstack-service/src/accessibility-handler.ts:
```diff
@@ -216,8 +243,26 @@ class _AccessibilityHandler {
+            if (!isPageOpened) {
+                return
+            }
             if (this._sessionId) {
```
this can be removed if we remove `isPageOpened` at whole?

nishath-bs replied:
Removed isOpened

kamal-kaur04 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -526,7 +588,66 @@ export const getA11yResults = async (browser: WebdriverIO.Browser, isBrowserStac
+        BStackLogger.debug('Performing scan before getting results summary')
+        await performA11yScan(isAppAutomate, browser, isBrowserStackSession, isAccessibility)
+        const apiUrl = `${APP_ALLY_ENDPOINT}/${APP_ALLY_ISSUES_ENDPOINT}`
+        let upperTimeLimit = Date.now() + 360000 // 30 seconds
```
Please handle this from backend :)

kamal-kaur04 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -526,7 +588,66 @@ export const getA11yResults = async (browser: WebdriverIO.Browser, isBrowserStac
+        BStackLogger.debug('Performing scan before getting results summary')
+        await performA11yScan(isAppAutomate, browser, isBrowserStackSession, isAccessibility)
+        const apiUrl = `${APP_ALLY_ENDPOINT}/${APP_ALLY_ISSUES_SUMMARY_ENDPOINT}`
+        const upperTimeLimit = Date.now() + 360000 // 30 seconds
```
360000 ms = 360 secs right? not 30 secs

sauravdas1997 replied:
we are polling for 6 minutes ? Is that intended ?
In certain cases the test runner can timeout, resulting in ingestor instability

nishath-bs replied:
We have changed it to 30 secs as default

kamal-kaur04 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -1359,3 +1480,68 @@ export const isValidCapsForHealing = (caps: { [key: string]: Options.Testrunner
+    message?: string; // Optional message for timeout cases
+  };
+
+async function pollApi(
```
```
  async function customPollApi(  url,
    params,
    headers,
    upperLimit,
    startTime = Date.now()) {
      return new Promise( async (resolve, reject) => {
        while(true) {
          try {
            params.timestamp = Date.now() / 1000
            BStackLogger.debug(`current timestamp ${params.timestamp}`)
            const response = await axios.get(url, { params, headers })
      
            // If the request succeeds (non-404), return the result
            resolve({
                data: response.data,
                headers: response.headers,
                message: 'Polling succeeded.',
            });
        } catch (error) {
          if (axios.isAxiosError(error) && error.response?.status === 404) {
              const nextPollTime = parseInt(error.response.headers.next_poll_time, 10) * 1000
              BStackLogger.debug(`timeInMillis ${nextPollTime}`)
              if (isNaN(nextPollTime)) {
                  BStackLogger.warn('Invalid or missing `nextPollTime` header. Stopping polling.')
                  reject({
                      data: null,
                      headers: error.response.headers,
                      message: 'Invalid timeInMillis header value. Polling stopped.',
                  })
              }
    
              const elapsedTime = nextPollTime - Date.now()
              BStackLogger.debug(`elapsedTime ${elapsedTime} timeInMillis ${nextPollTime} upperLimit ${upperLimit}`)
              // Stop polling if the upper time limit is reached
              if (nextPollTime > upperLimit) {
                  BStackLogger.warn('Polling stopped due to upper time limit.')
                  resolve({
                      data: null,
                      headers: error.response.headers,
                      message: 'Polling stopped due to upper time limit.',
                  })
              }
    
              BStackLogger.debug(`Polling again in ${elapsedTime}ms with params:`, params)
    
              // Wait for the specified time and poll again
              await new Promise(resolve => setTimeout(resolve, elapsedTime))
          } else if (axios.isAxiosError(error)) {
              reject({ data: {}, headers: {}, message: error?.response?.data.message });
          } else {
            BStackLogger.error('Unexpected error occurred:')
            reject({ data: {}, headers: {}, message: 'Unexpected error occurred.' });
          }
        }
        }

      });
    }
```

kamal-kaur04 replied:
shouldn't it be something like this?

nishath-bs replied:
I have handled it via recursive call and setTimeOut so the behaviour will be same

kamal-kaur04 commented on packages/wdio-browserstack-service/src/instrumentation/funnelInstrumentation.ts:
```diff
@@ -127,6 +127,9 @@ function buildEventData(eventType: string, config: BrowserStackConfig): any {
         eventProperties.isPercyAutoEnabled = config.isPercyAutoEnabled
         eventProperties.percyBuildId = config.percyBuildId
+        if (process.env[BSTACK_A11Y_POLLING_TIMEOUT]) {
+            eventProperties.allyPollingTimeout = process.env[BSTACK_A11Y_POLLING_TIMEOUT]
```
`pollingTimeout`, this is the name in Java.. So, let's keep consistent.

nishath-bs replied:
Done

nishath-bs committed:
changed prop name for instrumentation

nishath-bs committed:
log message changes

sauravdas1997 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -419,6 +427,23 @@ export const launchTestSession = o11yErrorHandler(async function launchTestSessi
+    BStackLogger.debug(`platformMeta ${JSON.stringify(platformMeta)}`)
+    try {
+        if (platformMeta?.platform_name && String(platformMeta?.platform_name).toLowerCase() === 'android') {
+            if (parseInt(platformMeta?.platform_version?.toString()) < 11) {
```
toString will result in `TypeError` if value is undefined

sauravdas1997 replied:
toString will result in TypeError if value is undefined, 
lets add proper handling for undefined case

sauravdas1997 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -526,7 +588,66 @@ export const getA11yResults = async (browser: WebdriverIO.Browser, isBrowserStac
+        const apiUrl = `${APP_ALLY_ENDPOINT}/${APP_ALLY_ISSUES_ENDPOINT}`
+        let upperTimeLimit = Date.now() + 300000 // 5 minutes
+        if (process.env[BSTACK_A11Y_POLLING_TIMEOUT]) {
+            upperTimeLimit = Date.now() + parseInt(process.env[BSTACK_A11Y_POLLING_TIMEOUT])
```
`process.env[BSTACK_A11Y_POLLING_TIMEOUT] ? Date.now() + parseInt(process.env[BSTACK_A11Y_POLLING_TIMEOUT]) : Date.now() + 300000`

sauravdas1997 commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -526,7 +588,66 @@ export const getA11yResults = async (browser: WebdriverIO.Browser, isBrowserStac
+        BStackLogger.debug('Performing scan before getting results summary')
+        await performA11yScan(isAppAutomate, browser, isBrowserStackSession, isAccessibility)
+        const apiUrl = `${APP_ALLY_ENDPOINT}/${APP_ALLY_ISSUES_ENDPOINT}`
+        let upperTimeLimit = Date.now() + 300000 // 5 minutes
```
5 mins is too high, test runner will timeout. Is this expected

nishath-bs committed:
review comment changes

dependabot[bot] committed:
bump the patch-deps-updates-v8 group across 1 directory with 18 updates (#14016)

* bump the patch-deps-updates-v8 group across 1 directory with 18 updates

---
updated-dependencies:
- dependency-name: "@preact/preset-vite"
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: "@types/archiver"
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: "@vitejs/plugin-react"
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: ejs-lint
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: husky
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: lerna
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: unzipper
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: csv-stringify
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: "@vitest/spy"
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: mlly
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: "@browserstack/ai-sdk-node"
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: "@percy/appium-app"
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: browserstack-local
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: "@vitest/snapshot"
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: chokidar
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: devtools-protocol
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: object-inspect
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
- dependency-name: testingbot-tunnel-launcher
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: patch-deps-updates-v8
...

Signed-off-by: dependabot[bot] <support@github.com>

* ignore deprecation errors

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Christian Bromann <git@bromann.dev>

nishath-bs committed:
removed isPageOpened

nishath-bs committed:
linting fixes

nishath-bs committed:
added math round off and moved logger inside scannerVersion check

sauravdas1997 commented:
LGTM

dbs-2001 committed:
add unit test for util file changes

dbs-2001 committed:
add test cases for app automate flow in accessibility handler

dbs-2001 committed:
add performa11yscan function testing for isappautomate true

dbs-2001 committed:
fix test cases in accessibility handle

dbs-2001 committed:
fix test cases in util

nishath-bs committed:
Trigger Build

nishath-bs committed:
linting fixes

nishath-bs committed:
added axios dependency

dbs-2001 committed:
test case fixed in getAppA11yResults

nishath-bs committed:
Trigger Build

nishath-bs committed:
Merge pull request #5 from nishath-bs/unit-test-mocha-cucumber-changes

add unit test for util file changes

dbs-2001 committed:
add new lines at last in test files

nishath-bs committed:
removed axios package

nishath-bs committed:
Update packages/wdio-browserstack-service/src/util.ts

Co-authored-by: Christian Bromann <git@bromann.dev>

nishath-bs committed:
compile issue fix

nishath-bs committed:
removed try catch and pollApi changes to without promise

nishath-bs committed:
changed unit test for got request module

dbs-2001 commented:
LGTM

dbs-2001 committed:
comment for got module usage

dbs-2001 committed:
remove commas

dbs-2001 committed:
fix ut

dbs-2001 committed:
remove format

rounak610 committed:
WDIO-V8 tests fixes (#14148)

* fixes in e2e mac tests

* fixes

* fixes

* fixes

* fixes

* fixes in cucumber

* fixes in cucumber tests

* commenting out step def in then.js

* revert changes

* fixes in lit.test.js

* fixes

* fixes

* fixes

* fixes

* fixes

* fixes

* fixes

* fixes

* fixes

* revert changes

* fixes

udarrr committed:
[WDIO8] fixed empty process.execArgv (#14150)

Co-authored-by: Christian Bromann <git@bromann.dev>

rounak610 committed:
fixes for percy in wdio-v8 (#14147)

Co-authored-by: Christian Bromann <git@bromann.dev>

sauravdas1997 committed:
Fix writable paths (#14052)

* fix: Handle paths for accessibility

* fix: EOF

* test: Add test cases

* refactor: EOF

---------

Co-authored-by: Christian Bromann <git@bromann.dev>

christian-bromann committed:
Merge branch 'v8' into app_a11y_assert_api
